src/struct.c

```diff
@@ -631,6 +631,52 @@ Dsymbol *AggregateDeclaration::searchCtor()
     return s;
 }
 
+/***************************************
+ * Return true if aggregate is an InputRange.
+ */
+bool AggregateDeclaration::isInputRange()
+{
+    return (this->search(Loc(), Id::Ffront) &&
```

Comment:
`front` and `empty` do not have to be functions, but indeed when they are functions, they must be nullary, and `front` must be of a non-void type. Strictly speaking, `empty` does not have to be of `bool` type but of a type castable to `bool`. `popFront` must be a nullary function.

Additionally, these properties do not have to be implemented as members, they can be implemented as free-standing property functions, which satisfy the concept through UFCS.

I don't know how foreach deals with it, but I'm guessing it uses error suppression, much like the `isInputRange` checker template itself.